### DISCUSSION TITLE
Service calls with return values for platform entity services

### DISCUSSION THREAD

**Discussion by eifinger** [home-assistant/architecture#948] [Entity Models] [answered]
In the discussion https://github.com/home-assistant/architecture/discussions/777 it was decided to add return values for services.
https://github.com/home-assistant/core/pull/94759 implemented this decision for the `EntityComponent` and the `calendar` platform. 

To enable more services to support return values we should add it to `EntityPlatform` as well.

**Pro:**

* Enables to get detailed travel time information , including the route https://github.com/home-assistant/core/pull/96141
* Enables services to get travel time information for supplied origin/destination without configuring it first
* Enables services for reverse geo-coding to return the resolved address from a set of coordinates. The address might be used to call another service e.g. weather service to get a weather forecast for the current location by city name.

**Con:**

* Concerns where raised in https://github.com/home-assistant/core/pull/96370

**Comment by allenporter** [comment]
I am not aware of additional concerns to address.  My impression is we already have a stance on the primary concern around matching multiple entities, unless there are more implementation details i'm missing. Prior discussions:
- The original architecture discussion high level proposal https://github.com/home-assistant/architecture/discussions/777 :
> For entity services: No return values when matching multiple entities
- In the [initial PR](https://github.com/home-assistant/core/pull/94401) we omitted from the original PR to reduce its scope while we overhauled and simplified the service call APIs and decided on the implementation detail
- In a followup PR the `entity_component` was updated when [adding calendar](https://github.com/home-assistant/core/pull/94759) which is an entity service enforces the original proposal that no return values are returned when matching multiple entities. The implementation detail is that we decided to raise when multiple entities are matched

To move this forward, I think we need the core architecture folks to weigh in on whether there is something else explicit we're missing here: Either additional concerns around platform entities missed from the original proposal, an implicit desire to support return values for multiple entities (where calendar needs an answer) or if we're on the same page with the prior work here.

**Comment by gjohansson-ST** [comment]
I assume this applies for https://github.com/home-assistant/core/pull/97280 as well then?

**Comment by Shulyaka** [comment]
I'd also like to have this feature enabled.
Just because a service call _might_ target multiple entities, it doesn't mean it _will_.
My use case: I want to create a service for a device, and I need any entity to identify that device. It feels more natural to use platform entity services for that because any entity would do.

So I suggest to start with allowing responses in platform entities when only one entity is targeted. It would be a caller's responsibility to ensure that. Then we can think of extending that into multiple entities.

**Reply by allenporter** [reply]
Just to clarify -- We've already decided how targets for multiple entities are handled when adding support in needed for calendar ([pr](https://github.com/home-assistant/core/pull/94759)).

What are the differences in concerns for platform entities vs the way it is already supported in `entity_component` to support calendar? As I mentioned in the other thread, I don't understand what else needs to be discussed here.

**Reply by gjohansson-ST** [reply]
I think (hope so) this discussion is to allow integrations to implement service calls with return values as well as entity components is already allowed to. Coming from Frenck's comment in https://github.com/home-assistant/core/pull/96370
Don't think multiple entities itself is any discussion that needs to be here.

It already works so it only needs to be allowed to do so.

**Reply by Shulyaka** [reply]
Yes. I apparently made a similar PR: https://github.com/home-assistant/core/pull/97719

**Comment by allenporter** [comment]
> I think (hope so) this discussion is to allow integrations to implement service calls with return values as well as entity components is already allowed to. Coming from Frenck's comment in [home-assistant/core#96370](https://github.com/home-assistant/core/pull/96370) Don't think multiple entities itself is any discussion that needs to be here.

Yes, my question is more specific: what additional discussion needs to happen? What's missing from the [original]( https://github.com/home-assistant/architecture/discussions/777) architecture discussion? What are the questions we need to answer or additional considerations we need to talk through for this case? IMO this has all been settled, so I can't tell if there is disagreement or misunderstanding.

**Reply by gjohansson-ST** [reply]
Given https://github.com/home-assistant/core/pull/97719 it doesn't sound like the multiple entities discussion is settled. 

Any other open I don't know about and think we just need to say yes or no to allow integrations to use it.

**Reply by allenporter** [reply]
The original PR we discussed that we didn't need it actually need that method yet and removed it from the scope. There was a discussion about possibly supporting this
https://github.com/home-assistant/core/pull/94401#discussion_r1228921697 and it consisted of a decision to revert that method, and look at in the next iteration.

In the next iteration, implemented one of the items in that comment thread and was the implementation from the original proposal: `For entity services: No return values when matching multiple entities` and it is implemented  https://github.com/home-assistant/core/pull/94759 (see [discussion](https://github.com/home-assistant/core/pull/94759#discussion_r1238221067) on specifics about return types of different functions) and the PR was approved by two core members.

The example in the documentation is an entity service:
https://www.home-assistant.io/docs/scripts/service-calls/#use-templates-to-handle-response-data and if you call the calendar list events service today and match multiple calendars it will fail: `Service call requested response data but matched more than one entity` This is the current behavior in Home Assistant for an entity service for calendar:
<img width="939" alt="Screenshot 2023-08-12 at 11 00 22 AM" src="https://github.com/home-assistant/architecture/assets/6026418/cc45ed2d-7997-4b6e-8157-80de182cc131">. My impression is that the weather forecast service is also an entity service ([pr](https://github.com/home-assistant/core/pull/97078))

I propose we preserve this behavior proposed in the original discussion that is currently implemented today, and extend to platform entity services, with the same checks implemented for component entity services.

If we want to support multiple entities in a single service call, some considerations we can continue to talk through:
- Two options for this I can think of for how to support this are: (1) Return a dict of entity id as the key to service response as the value as previously discussed or (2) Return a list of a dict with keys `entity_id` of the matched entity and `response` with the response. 
- We then have to find a balance for handling the simple single entity case with the multiple entity case. Or the fact that some services respond one way and entity services respond another way isn't great.  If we change the API to always support multiple entity service call responses, I think the case for a single entity more complex: I'm not a jinja expert at all, but i imagine to  always pick the first item in a list, or always iterate over items in the dictionary, or handling the case where it is possibly empty you're looking at things like this:`{{ responses[0]['response'] }}`, `{{ (response_data | items | first)[1] }}`. An option  is to have different response formats depending on if multiple entities are matched or if single entities are matched though that seems confusing and difficult to work with in a blueprint, or means you have to rewrite the script when moving from multiple entities to a single entity.
- An option could be to opt into support for matching multiple entities (e.g. say a script variable `target_response_variable`).
- Similar to the previous architecture discussion,  partial failure / partial success should not be supported across entities either.
- Do we have any motivating use cases in mind that we desire to support this? Perhaps that helps motivate any changes and/or apis. I currently do not have any supported use cases in mind.

An option is to revert the calendar list events and weather forecast service. Another is to move these to not be an entity service, though I don't know if that requires a breaking change.

**Reply by gjohansson-ST** [reply]
I agree with preserving it as is and expand the current implementation into platform entity services. 

Multiple entities should be a different discussion than this one or any features or improvements.

**Reply by frenck** [reply]
Removed the answer, as I thin not enough votes or activity took place.

Based on current knowledge and discussion I have seen in Discord, I really think we should consider different solutions.

With the experience now, and based on previous discussions I had with people (for example @ludeeus), I think we should explore a concept of returning values for the multiple targets.

This, imho, is not a different or separate discussion at all, as entity target are designed to be multiple from the start. Therefore we should at least consider handling them as such for responses as well.

I think we need to fully cover the multiple targets case, before we continue accepting anything at this point.

I will add this topic to a meeting with Paulus, Erik and Martin for upcoming week.

**Reply by allenporter** [reply]
> With the experience now, and based on previous discussions I had with people (for example @ludeeus), I think we should explore a concept of returning values for the multiple targets.
...
> I will add this topic to a meeting with Paulus, Erik and Martin for upcoming week.

Thank you for prioritizing the discussion. Just to be clear, i'm happy to keep the discussion moving and do the work to drive it forward. I put out some initial exploration/considerations above in the second half of the answer, and I was sincere about being happy to talk them through. If its more efficient to discuss in the meeting than this discussion thread thats fine with me.

**Reply by raman325** [reply]
@frenck were you able to make any progress on this? Just curious where we are at, it would be neat to have and I have a use case (zwave_js locks and a service to return user codes)

**Reply by frenck** [reply]
Yes! I have two paths to write out, however, didn't make it before I left on holiday and are now traveling too

**Reply by raman325** [reply]
enjoy and look forward to hearing more about it when you return!

**Comment by frenck** [comment]
Alright, so first of all, sorry for my late reply on this one. I've discussed this with Paulus, Erik, Martin & Robert a bit ago and this is a summary of where this can go.

There are three possible paths:

## 1. Allow service calls with responses only on a single entity

This means that after resolving the target (including area and/or device), if only a single target entity is left, we are able to return a response. In case a call returns multiple entities, we will raise.

**Pro**: The nice thing about this is that the response value is just as any non-entity service and thus maybe more familiar.

**Cons**: When targeting, for example, an area, or when using a template for the target; a change in the environment (e.g., new device added to an area), may cause anything that uses the service to fail all of a sudden. Thus may break automation and script in surprising ways.

## 2. Return a dictionary when multiple entities are target

When multiple entities are resolved, return a dictionary keyed with the entity ID. For example (in YAML data format for readability in the example):

```yaml
light.living_room:
 some_value: true
light.kitchen:
 some_value: false
```

However, when a single entity is targeted, return the normal version:

```yaml
some_value: true
```

**Pro**: The output is kinda more expected. When targeting a single entity you'll just get data from a single entity, when targeting multiple, you'll get multiple.

**Cons**: When targeting, for example, an area, or when using a template for the target; a change in the environment (e.g., new device added to an area) that cases the response to go from single to multiple results, may cause anything that uses the service to fail all of a sudden as the returned data format changes. This may break automation and script in surprising ways.

## 3. Always return a dictionary when using entity services

Similar to option 2, expect we will always return a dictionary:

```yaml
light.living_room:
 some_value: true
```

and

```yaml
light.living_room:
 some_value: true
light.kitchen:
 some_value: false
```

**Pro**: Consistency

**Cons**: A little harder to work with / process in templates.


# Suggestion

From these perspectives, we lean towards the solution in **option 3**.

The consistency (and thus avoiding breaking things all of a sudden later on), seems like an important value in this case. As it prevents surprises later on.

While it is a little less friendly to use, it will be consistent in the output provided, while also maintaining the target ability feature that entity service have.


../Frenck

**Reply by raman325** [reply]
FWIW as I read option 1 and 2 and the pros and cons, I settled on 3 before I read it 🙂 IMO consistency trumps every other pro, that's what makes it automateable.

The only thought I had on option 2 was that perhaps we can add some sort of template function/filter that allows you to pick the entity to filter on that abstracts any complexity away. The filter is smart enough to return the single entity payload under the entity id key, whether it is there (because of a multiple entity payload) or not (because of a single entity payload). This idea isn't fully formed but I think the larger point is that we can provide better ergonomics through templating where we see friction

**Reply by gjohansson-ST** [reply]
My vote is also on option 3 and would be very happy to see this getting implemented shortly.

**Reply by allenporter** [reply]
Appreciate you taking the time to follow up. This matches my take above in the previous answer (*If we want to support multiple entities in a single service call*) which is: if we want this a dictionary by entity is needed, and the down side is that templates are more complex. 

We'll need a migration plan for the existing entity services which I believe are just calendar and weather. My impression is these are entity services.

Aside: I hesitate to slow this down since i'd rather see the momentum continue :) though I still would like to see some motivating use cases spelled out. e.g. users want to merge weather forecasts. This will help as we consider usability improvements in the future.

**Reply by frenck** [reply]
> We'll need a migration plan for the existing entity services which I believe are just calendar and weather. My impression is these are entity services.

They are. I don't see a good migration path for them myself right now. It would be breaking at this point I guess.

**Reply by frenck** [reply, marked as answer]
Discussed this today again, this time include the context on the breaking change for the services. And this is what we ended up with:

We are going with [option 3 set out above](https://github.com/home-assistant/architecture/discussions/948#discussioncomment-7197201).

This means two services will be breaking: `weather.get_forecast` and `calendar.list_events`.
There is no way to ease this and the only solution we see is: 

- Introduce new services with a slightly different name, providing the new output from option 3.
- Keep the output of those as is, and deprecate them for 6 month. Raise an repair issue if they are used.

../Frenck

**Reply by eifinger** [reply]
Edit: Moved follow up question to new discussion: https://github.com/home-assistant/architecture/discussions/991
